### PR TITLE
Python: Add example and document for NIM Semantic kernel plugin  

### PR DESCRIPTION
**Pull request by josephwnv**
### Contribution Checklist

- [ ] The code builds clean without any errors or warnings
- [ ] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [ ] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

### CODE REVIEW

File: python/samples/demos/nvidia-nim/nvidia_nim_plugin.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    settings.function_call_behavior = FunctionCallBehavior.EnableFunctions(
+        auto_invoke=False, filters={"included_plugins": ["nllama3"]}
+    )
+    chat_history.add_user_message(
+        "What does nllama3 think about global warming?"
+    )
+
+    while True:
+        # The result is a list of ChatMessageContent objects, grab the first one
+        result = await chat.get_chat_message_contents(chat_history=chat_history, settings=settings, kernel=kernel)
```
Comment (eavanvalkenburg) [inline python/samples/demos/nvidia-nim/nvidia_nim_plugin.py:117]: you can also use `get_chat_message_content` then you do not get a list!

**Review by eavanvalkenburg — COMMENTED** [review]
nice one! small note

File: python/samples/demos/nvidia-nim/README.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
+# NVidia NIM Plugin
+
+NVidia Inference Microservice is fully optimized and has wide variety.  Perfect tools to enpower Copilot's semantic kernel.
```
Comment (TaoChenOSU) [inline python/samples/demos/nvidia-nim/README.md:3]:
```suggestion
NVidia Inference Microservice is fully optimized and has wide variety. Perfect tools to empower Copilot's semantic kernel.  
```

File: python/samples/demos/nvidia-nim/README.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
+# NVidia NIM Plugin
+
+NVidia Inference Microservice is fully optimized and has wide variety.  Perfect tools to enpower Copilot's semantic kernel.  
+
+This sample show how to encorperate NIM into semantic kernel.
+This sample is based on llama-3.1-8b-instruct:latest which is version 1.1.2 at this time.  Please check the the documentation of the NIM you plan to sue to see whethere there is any additional change you need to make.
```
Comment (TaoChenOSU) [inline python/samples/demos/nvidia-nim/README.md:6]:
```suggestion
This sample is based on llama-3.1-8b-instruct:latest which is version 1.1.2 at this time. Please check the documentation of the NIM you plan to use to see whether there is any additional change you need to make.
```

File: python/samples/demos/nvidia-nim/README.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+## Deploy NIM to Azure
+
+NIM can deploy to anyplace include but not limited to Azure ML. AKS and Azure VM.  Just do one of the following to prepare NIM endpoint for next step.
```
Comment (TaoChenOSU) [inline python/samples/demos/nvidia-nim/README.md:10]:
```suggestion
NIM can be deployed to anywhere including but not limited to Azure ML, AKS and Azure VM.  Just perform one of the following to prepare a NIM endpoint for next step.
```

File: python/samples/demos/nvidia-nim/README.md [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+1. **Azure ML Deployment:**
+
+ - Detail instruction can be found [here](https://github.com/NVIDIA/nim-deploy/tree/main/cloud-service-providers/azure/azureml)
```
Comment (TaoChenOSU) [inline python/samples/demos/nvidia-nim/README.md:14]:
```suggestion
 - Detail instructions can be found [here](https://github.com/NVIDIA/nim-deploy/tree/main/cloud-service-providers/azure/azureml)
```

File: python/samples/demos/nvidia-nim/README.md [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+        
+2. **Azure Kubernetes Service Deployment**
+   
+ - Detail instruction can be found [here](https://github.com/NVIDIA/nim-deploy/tree/main/cloud-service-providers/azure/aks)
```
Comment (TaoChenOSU) [inline python/samples/demos/nvidia-nim/README.md:18]:
```suggestion
 - Detail instructions can be found [here](https://github.com/NVIDIA/nim-deploy/tree/main/cloud-service-providers/azure/aks)
```

File: python/samples/demos/nvidia-nim/README.md [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+  
+3. **Azure VM Deployment**
+
+ - Create an Azure VM with 1x A100 GPU and NVidia AI Enterprise imGE 
+ - Follow the link [here](https://docs.nvidia.com/nim/large-language-models/latest/getting-started.html) to continue
+ - export the endpoint to public accessable.
```
Comment (TaoChenOSU) [inline python/samples/demos/nvidia-nim/README.md:24]:
```suggestion
 - Export the endpoint to public accessible.
```

File: python/samples/demos/nvidia-nim/README.md [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+## NVidia NIM Plugin 
+
+We use llama-3.1-8b-instruct as example.  We assume there is an expert called nllama3.  We create a plugin called nllama3 and a magic word called nllama3.  Any question asked nllama3 will redirect to this plugin and other questions will use default llm
```
Comment (TaoChenOSU) [inline python/samples/demos/nvidia-nim/README.md:28]:
```suggestion
We use llama-3.1-8b-instruct in this sample. We assume there is an expert called nllama3. We create a plugin called nllama3 and a magic word called nllama3. Any question asked to nllama3 will redirect to this plugin and other questions will use the default llm
```

File: python/samples/demos/nvidia-nim/README.md [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+ - Update the nim_url with the endpoint created in previous step.
+ - run nvidia_nim_plugin.py and see how it work.
```
Comment (TaoChenOSU) [inline python/samples/demos/nvidia-nim/README.md:31]:
```suggestion
 - Run nvidia_nim_plugin.py and see how it work.
```

File: python/samples/demos/nvidia-nim/nvidia_nim_plugin.py [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+nim_url = "http://0.0.0.0:8000/v1"
+
+class NLlama3Plugin:
+    """A sample plugin that provides response from NIM."""
+
+    @kernel_function(name="get_nllama3_opinion", description="Get the opinion of nllama3")
+    def get_nllama3_opinion(self, question: Annotated[str, "The input question"]) -> Annotated[str, "The output is a string"]:
+        
+        prompt = question.replace("nllama3", "you")
+        # Make sure model name match the model of NIM you deploy
```
Comment (TaoChenOSU) [inline python/samples/demos/nvidia-nim/nvidia_nim_plugin.py:31]:
```suggestion
        # Make sure model name matches the model of NIM you deploy
```

File: python/samples/demos/nvidia-nim/nvidia_nim_plugin.py [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+        client = OpenAI(base_url=nim_url, api_key="not-used")
```
Comment (TaoChenOSU) [inline python/samples/demos/nvidia-nim/nvidia_nim_plugin.py:32]: nit: What is the benefit of using a non-async client here?

File: python/samples/demos/nvidia-nim/nvidia_nim_plugin.py [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+    kernel.add_service(ai_service)
+
+    kernel.add_plugin(NLlama3Plugin(), plugin_name="nllama3")
+
+    # Example 1: Use automated function calling with a non-streaming prompt
+    print("========== Example 1: Use automated function calling with a non-streaming prompt ==========")
+    settings: OpenAIChatPromptExecutionSettings = kernel.get_prompt_execution_settings_from_service_id(
+        service_id=service_id
+    )
+    settings.function_call_behavior = FunctionCallBehavior.EnableFunctions(
```
Comment (TaoChenOSU) [inline python/samples/demos/nvidia-nim/nvidia_nim_plugin.py:67]: `FunctionCallBehavior` has been deprecated. We recommend to use `FunctionChoiceBehavior` instead.

**Comment by moonbox3** [comment]
Hi @josephwnv, we appreciate your contribution and would love to see this go into main; however, there are many comments that still need to be addressed. If you'd like to continue to support this, please pull latest code and work on addressing the comments, please. You may re-open the PR when you're ready. Thanks!

**Pull request closed without merging**